gapis/api/templates/vulkan_gfx_api_extras.tmpl

```diff
@@ -667,6 +667,45 @@ bool Vulkan::replayCreateVkDebugReportCallback(Stack* stack, bool pushReturn) {
     }
 }
 ¶
+bool Vulkan::replayWaitForFences(Stack* stack, bool pushReturn) {
+   auto expecteds = stack->pop<uint64_t*>();
+   auto fences = stack->pop<uint64_t*>();
+   auto count = stack->pop<uint64_t>();
+   auto device = static_cast<size_val>(stack->pop<size_val>());
+   if (stack->isValid()) {
+        GAPID_DEBUG("vkWaitForFences(%zu)", device);
+        uint32_t fence_status = VkResult::VK_TIMEOUT;
+        for (uint64_t i = 0; i < count; ++i) {
+          auto expected = expecteds[i];
+          if (expected == VkResult::VK_NOT_READY) {
+            continue;
+          }
+
+          if (expected == VkResult::VK_ERROR_DEVICE_LOST) {
+            fence_status = VkResult::VK_ERROR_DEVICE_LOST;
+            continue;
+          }
+
+          do {
+            fence_status = mVkDeviceFunctionStubs[device].vkGetFenceStatus(device, fences[i]);
+          } while (fence_status == VkResult::VK_NOT_READY);
+
+          if (fence_status == VkResult::VK_ERROR_DEVICE_LOST) {
+            break;
+          }
+        }
+
+        if (pushReturn) {
+            stack->push(fence_status);
```

Comment:
This seems incorrect: fence_status is the return value of vkGetFenceStatus(), and it is pushed here as the return value of vkWaitForFences().

Reply:
I have take a look at the code again and it is actually not wrong.

The VKResults can be returned from VkWaitForFences are:
- VK_SUCCESS
- VK_TIMEOUT
- VK_ERROR_OUT_OF_HOST_MEMORY
- VK_ERROR_OUT_OF_DEVICE_MEMORY
- VK_ERROR_DEVICE_LOST

and the VkResults that can be returned from VkGetFenceStatus are: 

- VK_SUCCESS (signaled)
- VK_NOT_READY (not signaled)
- VK_ERROR_DEVICE_LOST (error)

In trace time we keep the record of every trace time fence status. And what code does is:

If none of the fences signaled during trace time with no error -> Returns VK_TIMEOUT
if none of the fences signaled during trace time with any of them had error VK_ERROR_DEVICE_LOST -> Returns VK_ERROR_DEVICE_LOST

if any of the fence have been signaled and replay has no error-> Returns VK_SUCCESS
if any of the fence have been signaled and replay has error-> Returns VK_ERROR_DEVICE_LOST

So I think this is the expected behaviour.

As we do not actually call the VkWaitForFences ourselves, we cannot get the memory errors.

Reply:
This comment is invalid now because we have changed how we implement the vkWaitForFences. Please disregard.